source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1112 @@
+    }
+
+    // endpoint
+    struct aws_byte_cursor sts_cursor = aws_byte_cursor_from_c_str("sts.");
+    if (aws_byte_buf_init_copy_from_cursor(&parameter->endpoint, allocator, sts_cursor)) {
+        goto on_finish;
+    }
+    if (aws_byte_buf_append_dynamic(&parameter->endpoint, &region_cursor)) {
+        goto on_finish;
+    }
+    struct aws_byte_cursor amazonaws_cursor = aws_byte_cursor_from_c_str(".amazonaws.com");
+    if (aws_byte_buf_append_dynamic(&parameter->endpoint, &amazonaws_cursor)) {
+        goto on_finish;
+    }
+
+    if (aws_string_eq_c_str_ignore_case(region, "cn-north-1") ||
+        aws_string_eq_c_str_ignore_case(region, "cn-northwest-1")) {
+        struct aws_byte_cursor cn_cursor = aws_byte_cursor_from_c_str(".cn");
+        if (aws_byte_buf_append_dynamic(&parameter->endpoint, &cn_cursor)) {
+            goto on_finish;
+        }
+    }
+
+    // role_arn
+    if (!role_arn || !role_arn->len) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve role arn during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    struct aws_byte_cursor role_arn_cursor = aws_byte_cursor_from_string(role_arn);
+    if (aws_byte_buf_init_copy_from_cursor(&parameter->role_arn, allocator, role_arn_cursor)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve role arn during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    // token_file_path
+    if (!token_file_path || !token_file_path->len) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve token file path during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    struct aws_byte_cursor token_file_path_cursor = aws_byte_cursor_from_string(token_file_path);
+    if (aws_byte_buf_init_copy_from_cursor(&parameter->token_file_path, allocator, token_file_path_cursor)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to resolve token file path during sts web identity provider initialization.")
+        goto on_finish;
+    }
+
+    // role_session_name
+    if (role_session_name && role_session_name->len) {
+        struct aws_byte_cursor role_session_name_cursor = aws_byte_cursor_from_string(role_session_name);
+        if (aws_byte_buf_init_copy_from_cursor(&parameter->role_session_name, allocator, role_session_name_cursor)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Failed to resolve role session name during sts web identity provider initialization.")
+            goto on_finish;
+        }
+    } else {
+        struct aws_uuid uuid;
+
+        if (aws_uuid_init(&uuid)) {
+            AWS_LOGF_ERROR(AWS_LS_IO_PKI, "Failed to load an uuid during sts web identity provider initialization.");
```

Comment:
why is this a PKI error?

Reply:
Copied these code from somewhere else and forget to change the subject.